### PR TITLE
tcp: poll for writable during accept to see errors

### LINKED ISSUE #1112: TcpListener on iOS does not report bad file descriptor error
## Version

├── tokio v0.1.20
│   ├── tokio-codec v0.1.1
│   │   └── tokio-io v0.1.12
│   ├── tokio-current-thread v0.1.6
│   │   └── tokio-executor v0.1.7
│   ├── tokio-fs v0.1.6
│   │   └── tokio-threadpool v0.1.14
│   ├── tokio-reactor v0.1.9
│   │   └── tokio-sync v0.1.5
│   ├── tokio-tcp v0.1.3
│   ├── tokio-timer v0.2.11
│   ├── tokio-trace-core v0.1.0
│   ├── tokio-udp v0.1.3
│   │   ├── tokio-codec v0.1.1 (*)
│   └── tokio-uds v0.2.5

## Platform

iOS 12.2

## Description

I'm having trouble detecting when a listening TCP listening socket gets destroyed by iOS. I'm not sure whether it's a coding error, a bug in tokio, or a bug in mio. Perhaps someone can point me in the right direction!

The overall scenario in iOS:
1. Have app with a TCP listening socket
2. Go to home screen and background the app (threads get parked, sockets still alive at this point)
3. Lock device (sockets are killed by OS)
4. Unlock device and re-enter the running app

If I create a synchronous listener with `std::net::TcpListener` and loop on `incoming()`, I receive an error `Os { code: 9, kind: Other, message: "Bad file descriptor" }`, then an infinite series of `Os { code: 53, kind: ConnectionAborted, message: "Software caused connection abort" }` errors.

Using `tokio::net::TcpListener`, my stream task receives no values or errors to indicate that anything happened.

The async server, slightly modified from the tokio guide:
```rust
        let addr = "0.0.0.0:6142".parse().unwrap();
        let listener = tokio::net::TcpListener::bind(&addr).unwrap();
        let server = listener.incoming().for_each(|socket| {
            info!("New connection arrived on 6142");
            let (reader, writer) = socket.split();
            let amount = tokio::io::copy(reader, writer);

            let msg = amount.then(|result| {
                    match result {
                    Ok((amount, _, _)) => info!("wrote {} bytes", amount),
                    Err(e)             => info!("error: {}", e),
                }

                Ok(())
            });
            tokio::spawn(msg);
            Ok(())
        })
        .map_err(|err| {
            info!("Async accept error = {:?}", err);
        })
        .map(|_|{
            info!("Async ended normally");
        });
        executor.spawn(server);
        println!("Async TCP server running on 6142");
```

The sync server, slightly modified from the std docs:
```rust
        std::thread::spawn(move || {
            let listener = match std::net::TcpListener::bind("0.0.0.0:6143") {
                Ok(listener) => listener,
                Err(e) => {
                    error!("Error binding sync TCP listener: {}", e);
                    return;
                }
            };
            info!("Sync TCP server running on 6143");
            
            let mut err_count = 0;
            for stream in listener.incoming() {
                match stream {
                    Ok(_client) => {
                        info!("Sync TCP got a connection on 6143");
                    },
                    Err(e) => {
                        info!("Sync TCP incoming error: {:?}", e);
                        err_count += 1;
                        if err_count > 5 {
                            break;
                        }
                    }
                }
            }
            error!("Sync TCP shutting down!");
        });
```

Output when the app is first run and I make some connections:
```
Sync TCP got a connection on 6143
Sync TCP got a connection on 6143
Sync TCP got a connection on 6143
New connection arrived on 6142
wrote 4 bytes
New connection arrived on 6142
wrote 4 bytes
```

Log output after locking and unlocking:
```
Sync TCP incoming error: Os { code: 9, kind: Other, message: "Bad file descriptor" }
Sync TCP incoming error: Os { code: 53, kind: ConnectionAborted, message: "Software caused connection abort" }
Sync TCP incoming error: Os { code: 53, kind: ConnectionAborted, message: "Software caused connection abort" }
Sync TCP incoming error: Os { code: 53, kind: ConnectionAborted, message: "Software caused connection abort" }
Sync TCP incoming error: Os { code: 53, kind: ConnectionAborted, message: "Software caused connection abort" }
Sync TCP incoming error: Os { code: 53, kind: ConnectionAborted, message: "Software caused connection abort" }
Sync TCP shutting down!
```

### PR DESCRIPTION
**Pull request by thombles**
## Motivation

If the OS invalidates a TcpListener socket, this situation is detected by mio but not by tokio. As a result, the `incoming()` stream of remote clients hangs forever. (see #1112) 

Closes #1112

## Solution

If the listening socket is no longer valid, the error is reported by mio as a writable event. Therefore we must poll write and check to see if there is an error.

I can confirm that this change fixes the problem I experienced on iOS when locking/unlocking a device. I'm new to tokio's code, so please tell me if it should be handled differently. Also I'm not sure how to integration test this scenario. Do you have any suggestions for a good test?

### CODE REVIEW

**Comment by carllerche** [comment]
Sorry for the delay. Thanks for taking the time to debug and submit a PR.

Do you have a link to docs that describe receiving a writable event? Is that only for kqueue?

In what cases does the OS invalidate an OS listener?

**Comment by thombles** [comment]
> Do you have a link to docs that describe receiving a writable event? Is that only for kqueue?
> 
> In what cases does the OS invalidate an OS listener?

Great questions... I did a cursory search when I submitted the PR and came up blank. I've tried harder but I'm still not quite sure what's going on.

As far back as 1994, it was standard in BSD for sockets to be marked both readable and writable when an error occurred on a socket.
20190618_052529283_iOS

Supposedly kqueue follows the same semantics
20190618_053321089_iOS

The latest Darwin code I can find seems to support this definition of readability and writeability: https://github.com/apple/darwin-xnu/blob/a449c6a3b8014d9406c2ddbdc81795da24aa7443/bsd/kern/uipc_socket2.c#L2429-L2476

And yet, mio only gives a writable event. I annotated the mio code in `kqueue.rs` like so:

```rust
            if e.filter == libc::EVFILT_READ as Filter {
                info!("Found a readable event");
                event::kind_mut(&mut self.events[idx]).insert(Ready::readable());
            } else if e.filter == libc::EVFILT_WRITE as Filter {
                info!("Found a writable event");
                event::kind_mut(&mut self.events[idx]).insert(Ready::writable());
            }
```

...and in the iOS scenario (creating a TCP listener in an app, locking then unlocking the iPhone), it only mentions prints the "writable" message when the socket is found to be invalid on wake.

iOS is weird of course so I tried this on macOS too. Unfortunately I have no idea how to invalidate a socket on iOS. I tried binding specifically to my WLAN IP address and turning WiFi off, changing to a different static IP, etc., but the program simply sat there until I eventually got my original IP address back, at which point it started accepting connections again. ¯\\_(ツ)_/¯

So the question in my mind now is this: why are all the references telling me that a socket error should mark both readable and writable, yet a mio listener registered like this only gets a writable event when it fails?

```rust
            poll.register(&listener, Token(0), Ready::readable() | Ready::writable() | Ready::hup() | Ready::error(),
                        PollOpt::edge()).unwrap();
```

I'll keep digging as time allows, but if anyone has any insights I would love to know what's going on.

File: tokio-tcp/src/listener.rs [hunk 1]
```diff
@@ -178,6 +178,13 @@ impl TcpListener {
     /// # }
     /// ```
     pub fn poll_accept_std(&mut self) -> Poll<(net::TcpStream, SocketAddr), io::Error> {
+        // Error takes precedence, indicated by writable event on listening socket
```
Comment (carllerche) [inline tokio-tcp/src/listener.rs]: Is there a reason to put this first over trying to accept?

**Comment by carllerche** [comment]
What are some ways to invalidate a listener socket for tests?

**Comment by carllerche** [comment]
I dont' know what to do. I would like to get this merged in, but not having a test seems bad. I also don't know how to reproduce a TCP accept error in a test.

**Comment by thombles** [comment]
Sorry for leaving this hanging - I'm basically in the same boat. I know I need this but I don't quite know why and I don't know how to test it. I want to continue answering those questions but I haven't had more time yet. In a strict sense it's incomplete so we could close the PR for now, or leave it open in case someone who knows pops by and sees it.

**Comment by carllerche** [comment]
Well, it looks like this fixes a real issue, so if we can't come up w/ a test, we will move forward without one. I'm pinging some other people to see if they have thoughts.

**Comment by carllerche** [comment]
Ok, so EMFILE might be an error that we can use to simulate getting an error on the socket. You can try:

- create the listening socket.
- call socket() in a loop until you get EMFILE
- create one and then dup() it until EMFILE (Or just create one and then dup() it until EMFILE)
- And then try to connect that socket to the listening socket

That said, I am not sure this will result in a kqueue / epoll event. It is worth trying.

You mentioned socket *invalidation* which is not something I heard before. Do you mean this? https://developer.apple.com/documentation/corefoundation/1542010-cfsocketinvalidate If so, maybe we can have an os x only test that uses this API?

**Comment by thombles** [comment]
> Ok, so EMFILE might be an error that we can use to simulate getting an error on the socket.

This is a pretty nifty idea and I tried it out [over on another branch](https://github.com/thombles/tokio/commit/f803e903759684c6aa5c41b5a0b4898819ba5706). While it's not the most delightful code, it tests the FD scenario perfectly. The only trouble is it works with our without my change.

I added some logging to the listener poll and it shows this:

```
running 1 test
Read poll: Ok(NotReady)
Read poll: Ok(NotReady)
Read poll: Ok(NotReady)
Read poll: Ok(Ready(Readable))
test unix::listen_failure ... ok
```

So trying to `accept()` when we are out of fds on Mac triggers a Readable event, whereas my original scenario only produces Writable. That's rather annoying and I'll have to think about it some more.

> You mentioned socket _invalidation_ which is not something I heard before. Do you mean this?

No sorry I was using the term pretty loosely to mean that the OS was shutting down the listening socket. Unless I'm mistaken, that particular API you found is specific to CF wrapper cleanup rather than something I can do to a BSD socket.

File: tokio-tcp/src/listener.rs (hunk 1 above)
Comment (thombles) [reply tokio-tcp/src/listener.rs]: I don't think so - I was partly thinking I could maybe leave the `try_ready!` logic in place, partly that if a writable flag does in fact indicate more severe problems then maybe we should deal with that first. But if I can work out the other issues I'd be happy to put read and write on equal footing.

**Comment by carllerche** [comment]
@thombles in what situation does the OS forcibly shutdown the listener socket?

**Comment by thombles** [comment]
There's a full description on the issue (https://github.com/tokio-rs/tokio/issues/1112) but here's a summary:

1. Run an iOS app with Rust code and create a TCP listener
2. Lock the phone or tablet - connectivity is lost at this point, or a few seconds after
3. Unlock the device and resume the app if necessary - at this point the app becomes aware that the listening socket is dead. If there was a blocking accept() call it errors. In the case of mio/kqueue the writable flag is set and we find an error is present.

This is clearly impossible to unit test. It could be that iOS is just weird and we have to deal with it if we want to support that platform, but I can't rule out the possibility of some mio bug misusing kqueue. Unfortunately I find mio pretty difficult to follow so I think the next step is to make a minimal test case directly in C to confirm that the OS is behaving how I think it is.

**Comment by carllerche** [comment]
@thombles Can the issue be reproduced in the iOS simulator? If so, we could potentially write a test using that!

**Comment by carllerche** [comment]
@thombles since you verified this fixes your bug and #1112, lets just merge this. Could you add a big comment above the error check describing the use case and note that there currently is no test verifying that behavior?

**Comment by thombles** [comment]
It's a little unsatisfying but I'll go with that plan - it solves my problem. I think it's plausible that a unit test could be devised using the simulator. I can see references to unofficial APIs for locking the device programmatically. However I suspect that the resulting test infrastructure would be very large and fragile - I'm reluctant to attempt it.

**Comment by LucioFranco** [comment]
@thombles is this still something you need for tokio 0.1? I am going to close this but if you are still interested in pursing this for tokio 0.1 feel free to reopen. Sorry it took a long time to get to this. <3

**Comment by thombles** [comment]
Sorry to leave this hanging - technically I'm still making use of this fix (I've had this branch pinned for a while) but I'll be leaving tokio 0.1 behind very soon. I'll revisit it in 0.2, and if it remains so difficult to explain or test I'll probably resort to an application-level solution. Thanks!

**Comment by LucioFranco** [comment]
@thombles sounds good, ill be waiting for the PR when you have time :) Thanks!

**Pull request closed without merging**